Changes RestClient to allow for concurrent requests

RedworkDE opened the pull request:
# Details
- Changed Buckets from HashSet to ConcurrentDictionary (currently this can and will throw ocasionally)
- Removed the RequestSemaphore that serialized all requests
- To preserve the preemptive ratelimiting the following changes were made:
  - Decrement the remaining number of requests before the reqest is made, to allow this to happen atomically the Remaining property was changed to have a explicit backing field
  - Responses only update their ratelimit bucket when the response is for a newer interval than what the bucket currently points at. Otherwise this would push that remaining number of requests in that bucket back up when a response is received while other requests for that bucket are still pending
  - A ManualResetEvent has been added to signal that the global ratelimit has been hit and further requests should be delayed temporarily
- A HttpClientFactory has been added to allow for custom HttpClient implementation similar to how custom WebSocketClients and UdpClients are realized.

# Notes
- This should propably use a async implementation of the ManualResetEvent, but that would require additional dependencies and is only relevant in the case that the global ratelimit has been hit so I just used the framework implementation
  - the one big issue this could cause is that when spamming requests a deadlock could occour where all Threads from the TaskPool are blocked by the Wait, so that there are none left to set the event. A similar issue can occour now when a Exception is thrown while performing the request so that the Semaphore is never released
- This change will propably expose a couple of issues similar to #219 where the library assumes to be single-threaded while Tasks are in fact multi-threaded

Emzi0767 commented:
I spot a couple of issues, I'll do a proper review once I'm not on a mobile device.

Emzi0767 commented on DSharpPlus/Net/RateLimitBucket.cs:
```diff
@@ -41,7 +41,7 @@ public string BucketId
         /// Gets the number of uses left before pre-emptive rate limit is triggered.
         /// </summary>
-        public int Remaining { get; internal set; }
+        public int Remaining => _remaining;
```
If you're introducing a backing field to enable atomic writes, you should also make the property to perform an atomic read, preferably via `Volatile` or `Interlocked` classes.

RedworkDE replied:
All 32-bit reads are atomic (unless you are on some strange architecture or a 16-bit system but I am pretty sure .net does not run on such a platform). Making it a volatile read would be possible but the only place where it really matters is the check if there are requests available but that is implemented via the atomic decrement. The only place that uses that property now is the ToString method and there it probably won't matter if that is the most recent value as it will only be visible in the debugger

Emzi0767 commented on DSharpPlus/Net/RestClient.cs:
```diff
@@ -14,27 +15,35 @@
+    /// </summary>
+    /// <param name="messageHandler">Message handler to use for the new Http client instance.</param>
+    /// <returns>Constructed Http client implementation.</returns>
+    public delegate HttpClient HttpClientFactoryDelegate(HttpMessageHandler messageHandler);
```
See `HttpClientFactoryDelegate` comment in config class.

Emzi0767 commented on DSharpPlus/Net/RestClient.cs:
```diff
@@ -14,27 +15,35 @@
         private BaseDiscordClient Discord { get; }
         private HttpClient HttpClient { get; }
-        private HashSet<RateLimitBucket> Buckets { get; }
-        private SemaphoreSlim RequestSemaphore { get; }
```
Presumably this means you have implemented buckets to operate in an entirely atomic manner.

Emzi0767 commented on DSharpPlus/Net/RestClient.cs:
```diff
@@ -14,27 +15,35 @@
-        private HashSet<RateLimitBucket> Buckets { get; }
-        private SemaphoreSlim RequestSemaphore { get; }
+        private ConcurrentDictionary<string, RateLimitBucket> Buckets { get; }
+        private ManualResetEventSlim GlobalRateLimitEvent { get; }
```
If you're working in async code, you go async all the way. If you use synchronous waits, you will cause tasks to stall. Use a `TaskCompletionSource<T>` or another form of asynchronous waiting here.

RedworkDE replied:
The .met ManualResetEvent(Slim) has no async Wait method this would require using a library like AsyncEx or a custom implementation of such a event but since this only blocks when the global ratelimit has been hit i refrained from taking either action and went with the synchronous implementation included in the framework.

Emzi0767 replied:
Yes, but you can perform an async wait by using a `TaskCompletionSource<T>`, specifically awaiting its `.Task` property, and using `.SetResult(T)` to set the task to completed state.

RedworkDE replied:
Should be fixed, I added a AsyncManualResetEvent (I hope in the right place) for this

Emzi0767 commented on DSharpPlus/Net/RestClient.cs:
```diff
@@ -279,10 +282,16 @@ private void UpdateBucket(BaseRestRequest request, RestResponse response)
-            bucket.Remaining = int.Parse(usesleft, CultureInfo.InvariantCulture);
-            //bucket.Reset = servertime + resetdelta + difference;
-            bucket.Reset = clienttime + resetdelta;
+            // only update update the bucket values if this request was for a newer intervall than the one
```
Nitpick, but spelling: intervall -> interval.

Quahu replied:
also update update :robot:

Emzi0767 commented on DSharpPlus/Net/RestClient.cs:
```diff
@@ -14,27 +15,35 @@
         private HttpClient HttpClient { get; }
-        private HashSet<RateLimitBucket> Buckets { get; }
-        private SemaphoreSlim RequestSemaphore { get; }
+        private ConcurrentDictionary<string, RateLimitBucket> Buckets { get; }
```
You're saying this has the potential to throw, and throwing can cause deadlocks. If you're going with this approach, I would request that you make sure nothing around this throws.

RedworkDE replied:
I meant that the current implementation using the HashSet can throw when multiple threads make requests and the set is modified while the other thread still searches for its Bucket using the ConcurrentDictionary avoids this.

Emzi0767 replied:
A'ight, that's fair.

RedworkDE commented on DSharpPlus/DiscordConfiguration.cs:
```diff
@@ -99,6 +101,24 @@ public string Token
+        /// <para>Use <see cref="HttpClient(HttpMessageHandler)"/> and equivalents on other implementations to switch out client implementations.</para>
+        /// <para>Defaults to <see cref="HttpClient(HttpMessageHandler)"/>.</para>
+        /// </summary>
+        public HttpClientFactoryDelegate HttpClientFactory
```
I have a use case where I have to use a custom HttpClient (actually replacing the HttpMessageHandler would probably be better) but I can see why this doesn't really fit in this library so i'll remove that again and use reflection in my code.

Emzi0767 commented:
A'ight, I'll test these later on. If everything works fine, I will merge.

RedworkDE commented:
Preemptive ratelimiting does not work in 2 situations:
1. Sending more than ratelimit requests before the first one finishes
2. Sending more than 2x ratelimit requests in one ratelimit interval
I need to think about how/if this is solvable without hard coding the limits or falling back to sequential requests for a route.

Emzi0767 commented:
Yeah this is why we (and everyone else) does it by using a semaphore.

Emzi0767 commented:
Another issue is that if the ratelimit handler doesn't work 100% it could lead to library's users getting API-banned.

RedworkDE commented:
@Emzi0767 Do you happen to know if it is valid to assume if it is valid to assume that

a) method + route determines the maximum number of requests in a rate-limit interval?
Example: You can always post 5 message in a channel independently of the channel in question

b) the maximum number of requests in a rate-limit interval does not change for a bucket.
Example: Once you know that you can send X messages in a channel that number X does not suddenly change.

It is not too bad if a) is not valid this just means that before sending the second message in a bucket we have to wait until the first request succeeded.
If b) is not valid there are two ways of dealing with that: either accept that it will take one interval for the change to take effect meaning that we could have one or two requests that get rate-limited by the server because we were still using the old limit or alternatively do the same as for a) but for every rate-limit interval.

Emzi0767 commented:
1. Method + route is not a valid way of determining a bucket. It's method + route + major parameters.
2. Maximum shouldn't change.

The idea to wait per bucket instead of globally is OK, provided that all below requirements are met:
- No 429s are triggered as a result of failed ratelimiting.
- No deadlocks occur as a result of failures or such.
- Global ratelimit is handled properly.
- The bucket overrides work properly.

RedworkDE commented:
Ok, everything should be working now:
- Requests in different buckets will not influence each other unless the global rate limit has been triggered
- Requests in a bucket will be executed one after another until we get rate limit data for that bucket, so that only one request can hit the limit for each interval
- If the limit is knows the appropriate number of requests can happen concurrently, excess requests will still be blocked until the next interval
  - If this happens pretty much all the excess requests will log a warning for every interval they got blocked
- I fixed a bug causing the sharded client to ignore the proxy setting
- I added logging for uncaught exception that happen in execute request
- The only limitation that this has is that if the global rate limit has been hit, all currently executing requests will hit it as well but there is nothing that can be done about that since they are already on the wire

RedworkDE commented:
That timeout can probably cause all kinds of library methods to throw TaskCanceledExceptions (or similar), but I think this is still the best way of handling this. If a request did not finish after 10 seconds it is definitely dead but we can't just ignore it and return something (we have no response to return), we cannot resend the request (we don't know if the server processed the request or not and if it is safe to just resend it) and we can't silently drop it (this could cause the user to lock up as well) so passing the exception to the user it is.

This event should be really rare anyways, unless discord has an outage or the user is is on a unstable internet connection, so it shouldn't be too much of a problem.

Emzi0767 commented:
I'll merge this into a separate branch later; I want to do some thorough testing myself first.

RedworkDE commented:
One more update: I am using this since 3 months in prod without any issues, so i am going to assume this code to be bug-free™. If anybody wants to use this i will keep my fork up-to-date with this one, but note that i am going to keep rebasing this commit so developing something based on this is probably a bad idea until this is merged.
